pkg/runtime/runtime.go

```diff
@@ -0,0 +1,269 @@
+// Copyright 2021 The Codefresh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/codefresh-io/cli-v2/pkg/log"
+	"github.com/codefresh-io/cli-v2/pkg/store"
+
+	"github.com/Masterminds/semver/v3"
+	apcmd "github.com/argoproj-labs/argocd-autopilot/cmd/commands"
+	"github.com/argoproj-labs/argocd-autopilot/pkg/application"
+	"github.com/argoproj-labs/argocd-autopilot/pkg/fs"
+	"github.com/argoproj-labs/argocd-autopilot/pkg/git"
+	"github.com/argoproj-labs/argocd-autopilot/pkg/kube"
+	apstore "github.com/argoproj-labs/argocd-autopilot/pkg/store"
+	"github.com/ghodss/yaml"
+	billyUtils "github.com/go-git/go-billy/v5/util"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kusttypes "sigs.k8s.io/kustomize/api/types"
+)
+
+type (
+	Runtime struct {
+		metav1.TypeMeta   `json:",inline"`
+		metav1.ObjectMeta `json:"metadata"`
+
+		Spec RuntimeSpec `json:"spec"`
+	}
+
+	RuntimeSpec struct {
+		DefVersion         *semver.Version `json:"defVersion"`
+		Version            *semver.Version `json:"version"`
+		BootstrapSpecifier string          `json:"bootstrapSpecifier"`
+		Components         []AppDef        `json:"components"`
+	}
+
+	AppDef struct {
+		Name string `json:"name"`
+		Type string `json:"type"`
+		URL  string `json:"url"`
+		Wait bool   `json:"wait"`
+	}
+)
+
+func Download(version *semver.Version, name string) (*Runtime, error) {
+	var (
+		body []byte
+		err  error
+	)
+
+	if strings.HasPrefix(store.RuntimeDefURL, "http") {
+		urlObj, err := url.Parse(store.RuntimeDefURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed parsing url: %w", err)
+		}
+
+		if urlObj.Query().Get("ref") == "" {
+			urlObj.Query().Set("ref", version.String())
```

Comment:
should check wether version was set, it can be nil

Reply:
If the user doesn't supply a version, the command puts in the current binary version as the version. So when we get to here, it will always contain a valid value.